tests/msc2403_test.go

```diff
@@ -0,0 +1,281 @@
+	})
+
+	// Test knocking between two users on the same homeserver
+	knockingBetweenTwoUsersTest(t, roomIDOne, alice, bob, false)
+
+	// Create a room for alice and charlie to test knocking with
+	roomIDTwo := alice.CreateRoom(t, struct {
+		Preset      string `json:"preset"`
+		RoomVersion string `json:"room_version"`
+	}{
+		"private_chat",      // Set to private in order to get an invite-only room
+		"xyz.amorgan.knock", // Room version required for knocking. TODO: Remove when knocking is in a stable room version
+	})
+
+	// Test knocking between two users, each on a separate homeserver
+	knockingBetweenTwoUsersTest(t, roomIDTwo, alice, charlie, true)
+}
+
+func knockingBetweenTwoUsersTest(t *testing.T, roomID string, inRoomUser, knockingUser *client.CSAPI, federation bool) {
+	knockingUserID := knockingUser.UserID
+
+	t.Run("Knocking on a room with a join rule other than 'knock' should fail", func(t *testing.T) {
+		knockOnRoomWithStatus(t, knockingUser, roomID, "Can I knock anyways?", []string{"hs1"}, 403)
+	})
+
+	t.Run("Change the join rule of a room from 'invite' to 'xyz.amorgan.knock'", func(t *testing.T) {
+		inRoomUser.MustDo(
+			t,
+			"PUT",
+			[]string{"_matrix", "client", "r0", "rooms", roomID, "state", "m.room.join_rules", ""},
+			struct {
+				JoinRule string `json:"join_rule"`
+			}{
+				"xyz.amorgan.knock",
+			},
+		)
+	})
+
+	t.Run("Attempting to join a room with join rule 'xyz.amorgan.knock' without an invite should fail", func(t *testing.T) {
+		// Set server_name so we can find rooms via ID over federation
+		query := url.Values{
+			"server_name": []string{"hs1"},
+		}
+
+		knockingUser.MustDoWithStatusRaw(
+			t,
+			"POST",
+			[]string{"_matrix", "client", "r0", "join", roomID},
+			[]byte("{}"),
+			"application/json",
+			query,
+			403,
+		)
+	})
+
+	t.Run("Knocking on a room with join rule 'xyz.amorgan.xyz' should succeed", func(t *testing.T) {
+		knockOnRoom(t, knockingUser, roomID, testKnockReason, []string{"hs1"})
+	})
+
+	t.Run("A user that has already knocked cannot immediately knock again on the same room", func(t *testing.T) {
+		knockOnRoomWithStatus(t, knockingUser, roomID, "I really like knock knock jokes", []string{"hs1"}, 403)
+	})
+
+	// Now that the user has knocked on the room successfully, let's check that everything went right
+
+	t.Run("parallel", func(t *testing.T) {
+		t.Run("Users in the room see a user's membership update when they knock", func(t *testing.T) {
+			t.Parallel()
+			inRoomUser.SyncUntilTimelineHas(t, roomID, func(ev gjson.Result) bool {
+				if ev.Get("type").Str != "m.room.member" || ev.Get("sender").Str != knockingUserID {
+					return false
+				}
+				must.EqualStr(t, ev.Get("content").Get("reason").Str, testKnockReason, "incorrect reason for knock")
+				must.EqualStr(t, ev.Get("content").Get("membership").Str, "xyz.amorgan.knock", "incorrect membership for knocking user")
```

Comment:
I would make `const constKnock = "xyz.amorgan.knock"` and use that, which will also help for reducing the number of changes when it lands.